File: languages/at_archer/at_archer.rs
use proconio::input;
 
fn main() {
    input! {
        n: i64,
        m: i64,
        d: i64,
        r: [i64; m as usize + 1],
        s: [i64; m as usize],
    }
    let answer = solve(n, m, d, r, s);
    println!("{}", answer);
}

fn solve(n: i64, m: i64, d: i64, r: Vec<i64>, mut s: Vec<i64>) -> i64 {
    let a = -(n * d / 2);
    let mut tmp_vec: Vec<i64> = r.iter().rev().map(|x| -x).collect::<Vec<i64>>();
    tmp_vec = tmp_vec[..m as usize].to_vec();
    tmp_vec.append(&mut r.iter().map(|x| x + 1).collect::<Vec<i64>>()[1..].to_vec());
    let l = tmp_vec[..tmp_vec.len() - 1].to_vec();
    let r_vec = tmp_vec[1..].to_vec();
    let mut s_vec: Vec<i64> = Vec::new();
    for s0 in s.iter().rev() {
        s_vec.push(*s0);
    }
    s_vec.append(&mut s[1..].to_vec());
    let mut imos = vec![0; d as usize + 1];
    for i in 0..(2 * m as usize - 1) {
        if l[i] - a >= n * d {
            imos[0] -= n * s_vec[i];
            imos[d as usize] += n * s_vec[i];
        } else if l[i] - a >= 1 {
            imos[0] -= (l[i] - a + d) / d * s_vec[i];
            imos[((l[i] - a) % d) as usize] += s_vec[i];
            imos[d as usize] += (l[i] - a) / d * s_vec[i];
        }
        if r_vec[i] - a >= n * d {
            imos[0] += n * s_vec[i];
            imos[d as usize] -= n * s_vec[i];
        } else if r_vec[i] - a >= 1 {
            imos[0] += (r_vec[i] - a + d) / d * s_vec[i];
            imos[((r_vec[i] - a) % d) as usize] -= s_vec[i];
            imos[d as usize] -= (r_vec[i] - a) / d * s_vec[i];
        }
    }
    let mut val: i64 = 0;
    let mut score: i64 = 0;
    for i in 0..d as usize {
        val += imos[i];
        score = std::cmp::max(score, val);
    }
    return score;
}

File: languages/at_archer/at_archer.cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <iterator>
using namespace std;

long long solve(long long n, int m, long long d, const vector<long long>& r, const vector<long long>& s) {
	long long a = -(n * d / 2);
    vector<long long> tmp_vec1, tmp_vec2, l, r_vec, s_vec;
    reverse_copy(r.begin() + 1, r.end(), back_inserter(tmp_vec1));
    for_each(tmp_vec1.begin(), tmp_vec1.end(), [](long long& element) {
        element *= -1;
    });
    copy(r.begin() + 1, r.end(), back_inserter(tmp_vec2));
    for_each(tmp_vec2.begin(), tmp_vec2.end(), [](long long& element) {
        ++element;
    });
    copy(tmp_vec2.begin(), tmp_vec2.end(), back_inserter(tmp_vec1));
    copy(tmp_vec1.begin(), tmp_vec1.end() - 1, back_inserter(l));
    copy(tmp_vec1.begin() + 1, tmp_vec1.end(), back_inserter(r_vec));
    reverse_copy(s.begin(), s.end(), back_inserter(s_vec));
    copy(s.begin() + 1, s.end(), back_inserter(s_vec));

	vector<long long> imos(d + 1);
	for (int i = 0; i < 2 * m - 1; ++i) {
		if (l[i] - a >= n * d) {
			imos[0] -= n * s_vec[i];
			imos[d] += n * s_vec[i];
		}
		else if (l[i] - a >= 1) {
			imos[0] -= (l[i] - a + d) / d * s_vec[i];
			imos[(l[i] - a) % d] += s_vec[i];
			imos[d] += (l[i] - a) / d * s_vec[i];
		}
		if (r_vec[i] - a >= n * d) {
			imos[0] += n * s_vec[i];
			imos[d] -= n * s_vec[i];
		}
		else if (r_vec[i] - a >= 1) {
			imos[0] += (r_vec[i] - a + d) / d * s_vec[i];
			imos[(r_vec[i] - a) % d] -= s_vec[i];
			imos[d] -= (r_vec[i] - a) / d * s_vec[i];
		}
	}
	long long val = 0, score = 0;
	for (int i = 0; i < d; ++i) {
		val += imos[i];
		score = max(score, val);
	}
	return score;
}

int main() {
	long long n, m, d;
    cin >> n >> m >> d;
	vector<long long> r(m + 1), s(m);
	for (int i = 0; i < m + 1; ++i) {
		cin >> r[i];
	}
	for (int i = 0; i < m; ++i) {
		cin >> s[i];
	}
	long long answer = solve(n, m, d, r, s);
	cout << answer << endl;
	return 0;
}
